### PR TITLE
Bugfix: Clear DoFHandler in Triangulation create signal.

### LINKED ISSUE #11622: GridGenerator::merge_triangulations fails when input=output
This is extracted from a program by @juleoc02 and fails only with current master but not 9.2:
```
#include <deal.II/grid/tria.h>
#include <deal.II/grid/grid_generator.h>
#include <deal.II/dofs/dof_handler.h>

namespace SAND {
  using namespace dealii;

  template<int dim>
  class SANDTopOpt {
  public:
    SANDTopOpt();

    Triangulation<dim> triangulation;
    DoFHandler<dim> dof_handler;
  };

  template<int dim>
  SANDTopOpt<dim>::SANDTopOpt()
  :
  dof_handler(triangulation)
  {
    /*Make a square*/
    Triangulation<dim> triangulation_temp;
    Point<dim> point_1, point_2;
    point_1(0) = 0;
    point_1(1) = 0;
    point_2(0) = 1;
    point_2(1) = 1;
    GridGenerator::hyper_rectangle(triangulation, point_1, point_2);

    triangulation_temp.clear();
    point_1(0) = 1;
    point_2(0) = 2;
    GridGenerator::hyper_rectangle(triangulation_temp, point_1, point_2);
    /*glue squares together*/
    GridGenerator::merge_triangulations(triangulation_temp,
                                        triangulation, triangulation);
  }

} // namespace SAND

int
main() {
  SAND::SANDTopOpt<2> elastic_problem_2d;
}
```

It fails like this:
```
--------------------------------------------------------
An error occurred in line <3150> of file </home/bangerth/p/deal.II/1/dealii/source/dofs/dof_handler.cc> in function
    void dealii::DoFHandler<dim, spacedim>::create_active_fe_table() [with int dim = 2; int spacedim = 2]
The violated condition was: 
    this->hp_cell_active_fe_indices[level].size() == this->tria->n_raw_cells(level) && this->hp_cell_future_fe_indices[level].size() == this->tria->n_raw_cells(level)
Additional information: 
    This exception -- which is used in many places in the library -- usually indicates that some condition which the author of the code thought must be satisfied at a certain point in an algorithm, is not fulfilled. An example would be that the first part of an algorithm sorts elements of an array in ascending order, and a second part of the algorithm later encounters an element that is not larger than the previous one.

There is usually not very much you can do if you encounter such an exception since it indicates an error in deal.II, not in your own program. Try to come up with the smallest possible program that still demonstrates the error and contact the deal.II mailing lists with it to obtain help.

Stacktrace:
-----------
#0  /home/bangerth/p/deal.II/1/install/lib/libdeal_II.g.so.9.3.0-pre: dealii::DoFHandler<2, 2>::create_active_fe_table()
#1  /home/bangerth/p/deal.II/1/install/lib/libdeal_II.g.so.9.3.0-pre: dealii::DoFHandler<2, 2>::connect_to_triangulation_signals()::{lambda()#1}::operator()() const
```

I believe that the issue is that input triangulation = output triangulation.

@marcfehling -- any idea for a quick fix?

### PR DESCRIPTION
**Pull request by marcfehling**
Fixes #11622.

We forgot to clear the `DoFHandler` object in the `create` signal of a `Triangulation`.

### CODE REVIEW

**Review by peterrum — APPROVED** [review]

**Comment by nfehn** [comment]
maybe need to correct or clarify documentation, see issue #11622

File: include/deal.II/grid/grid_generator.h [hunk 1]
```diff
@@ -1577,6 +1577,12 @@ namespace GridGenerator
    * @note For a related operation on refined meshes when both meshes are
    * derived from the same coarse mesh, see
    * GridGenerator::create_union_triangulation().
+   *
+   * @note By internally calling Triangulation::create_create_triangulation(),
+   * this function triggers the "create" signal after doing its work. See
```
Comment (masterleinad) [inline include/deal.II/grid/grid_generator.h:1582]:
```suggestion
   * @note This function triggers the "create" signal after doing its work. See
```
We don't need to document what we do internally which might be subject to change.

Would you mind adding that it is legal for one of the input triangulations to be equal to `result` in response to https://github.com/dealii/dealii/issues/11622?

**Review by masterleinad — APPROVED** [review]
Looks good to me.

File: include/deal.II/grid/grid_generator.h [hunk 2]
```diff
@@ -1523,6 +1523,9 @@ namespace GridGenerator
    * the triangulation that contains the cells of both triangulation and store
    * it in the third parameter. Previous content of @p result will be deleted.
    *
+   * One of the two input triangulations can also be the @p result triangulation.
+   * Beware that this triangulation will be overwritten.
```
Comment (bangerth) [inline include/deal.II/grid/grid_generator.h:1527]: I think the second sentence is implied :-) If it's the result, *of course* its contents will be overwritten.

File: include/deal.II/grid/grid_generator.h [hunk 3]
```diff
@@ -1577,6 +1580,12 @@ namespace GridGenerator
    * @note For a related operation on refined meshes when both meshes are
    * derived from the same coarse mesh, see
    * GridGenerator::create_union_triangulation().
+   *
+   * @note This function triggers the "create" signal on the @p result
```
Comment (bangerth) [inline include/deal.II/grid/grid_generator.h:1584]: I'd just drop the note. The details aren't important.

File: source/dofs/dof_handler.cc [hunk 4]
```diff
@@ -3043,7 +3043,7 @@ DoFHandler<dim, spacedim>::connect_to_triangulation_signals()
 {
   // connect functions to signals of the underlying triangulation
   this->tria_listeners.push_back(this->tria->signals.create.connect(
-    [this]() { this->create_active_fe_table(); }));
+    [this]() { this->reinit(*(this->tria)); }));
```
Comment (bangerth) [inline source/dofs/dof_handler.cc:3046]: Minimal patch :-)

**Pull request merged**